### PR TITLE
Pims 976

### PR DESCRIPTION
**Pull request by devinleighsmith**

### CODE REVIEW

File: backend/api/Areas/Admin/Controllers/UserController.cs [hunk 1]
```diff
@@ -153,11 +153,25 @@ public IActionResult DeleteUser([FromBody] Model.UserModel model)
             return new JsonResult(model);
         }
 
+        /// <summary>
+        /// Update an access request, generally to grant/deny it.
+        /// </summary>
+        /// <returns></returns>
+        [HttpPost("/api/admin/access/request/")]
```
Comment (Fosol) [inline backend/api/Areas/Admin/Controllers/UserController.cs]: This should be a PUT as it's for updating not creating a new one.

File: backend/dal/Services/Admin/Concrete/UserService.cs [hunk 2]
```diff
@@ -150,6 +150,65 @@ public override void Remove(User entity)
...
+        /// Updates the specified access request in the datasource. if the request is granted, update the associated user as well.
+        /// </summary>
+        /// <param name="entity"></param>
+        /// <exception type="KeyNotFoundException">Entity does not exist in the datasource.</exception>
+        /// <returns></returns>
+        public AccessRequest Update(AccessRequest entity)
+        {
+            if(entity.UserId == null)
+            {
+                throw new ArgumentNullException("userId");
```
Comment (Fosol) [inline backend/dal/Services/Admin/Concrete/UserService.cs]: `nameof(entity.UserId)`

File: backend/dal/Services/Admin/Concrete/UserService.cs [hunk 3]
```diff
@@ -150,6 +150,65 @@ public override void Remove(User entity)
...
+            }
+            entity.ThrowIfNull(nameof(entity));
```
Comment (Fosol) [inline backend/dal/Services/Admin/Concrete/UserService.cs]: If `entity` is null then this line needs to be first.

File: backend/dal/Services/Admin/Concrete/UserService.cs [hunk 4]
```diff
@@ -150,6 +150,65 @@ public override void Remove(User entity)
...
+            this.User.ThrowIfNotAuthorized(Permissions.SystemAdmin, Permissions.AgencyAdmin);
```
Comment (Fosol) [inline backend/dal/Services/Admin/Concrete/UserService.cs]: I would change this to just `Permissions.AdminUsers`

File: backend/dal/Services/Admin/Concrete/UserService.cs [hunk 5]
```diff
@@ -150,6 +150,65 @@ public override void Remove(User entity)
...
+
+            var accessRequest = this.Context.AccessRequests
+                .Include(p => p.Agencies)
+                .ThenInclude(p => p.Agency)
+                .Include(p => p.Roles)
+                .ThenInclude(p => p.Role)
+                .Include(p => p.User)
+                .FirstOrDefault(a => a.Id == entity.Id) ?? throw new KeyNotFoundException();
+            if(!accessRequest.IsGranted != true && entity.IsGranted == true && accessRequest.UserId.HasValue)
```
Comment (Fosol) [inline backend/dal/Services/Admin/Concrete/UserService.cs]: Any reason for the double negative?  `accessRequest.IsGranted == true`.  No reason to double check if `accessRequest.UserId.HasValue` as you've checked earlier already.

File: backend/dal/Services/Admin/Concrete/UserService.cs [hunk 6]
```diff
@@ -150,6 +150,65 @@ public override void Remove(User entity)
...
+            {
+                User user = Get(accessRequest.UserId.Value);
+                entity.Agencies.ForEach((accessRequestAgency) =>
+                {
+                    if (!user.Agencies.Select(a => a.AgencyId).Contains(accessRequestAgency.AgencyId))
```
Comment (Fosol) [inline backend/dal/Services/Admin/Concrete/UserService.cs]: Change to `!user.Agencies.Any(a => a.AgencyId == accessRequestAgency.AgencyId)`

File: backend/dal/Services/Admin/Concrete/UserService.cs [hunk 7]
```diff
@@ -150,6 +150,65 @@ public override void Remove(User entity)
...
+                    {
+                        user.Agencies.Add(new UserAgency()
+                        {
+                            User = user,
+                            Agency = this.Context.Agencies.Find(accessRequestAgency.AgencyId)
```
Comment (Fosol) [inline backend/dal/Services/Admin/Concrete/UserService.cs]: No real value in fetching the agency here.  Just set the `AgencyId=accessRequestAgency.AgencyId`.  The result should be the same with one less trip to the DB.

File: backend/dal/Services/Admin/Concrete/UserService.cs [hunk 8]
```diff
@@ -150,6 +150,65 @@ public override void Remove(User entity)
...
+                        });
+                    }
+                });
+                entity.Roles.ForEach((accessRequestRole) =>
+                {
+                    if (!user.Roles.Select(r => r.RoleId).Contains(accessRequestRole.RoleId))
```
Comment (Fosol) [inline backend/dal/Services/Admin/Concrete/UserService.cs]: Change to `!user.Roles.Any(r => r.RoleId == accessRequestRole.RoleId)`

File: backend/dal/Services/Admin/Concrete/UserService.cs [hunk 9]
```diff
@@ -150,6 +150,65 @@ public override void Remove(User entity)
...
+                    {
+                        user.Roles.Add(new UserRole()
+                        {
+                            User = user,
+                            Role = this.Context.Roles.Find(accessRequestRole.RoleId)
```
Comment (Fosol) [inline backend/dal/Services/Admin/Concrete/UserService.cs]: No value in fetching the role here.  Just set the `RoleId=accessRequestAgency.RoleId`.  The result should be the same with one less trip to the DB.

File: backend/dal/Services/Admin/Concrete/UserService.cs [hunk 10]
```diff
@@ -150,6 +150,65 @@ public override void Remove(User entity)
...
+                        });
+                    }
+                });
+                Update(user);
```
Comment (Fosol) [inline backend/dal/Services/Admin/Concrete/UserService.cs]: This update to the user will occur in its own transaction.  You need to use `UpdateOne(user)`.  

Think I need a better naming convention for those methods...

File: backend/dal/Services/Admin/Concrete/UserService.cs [hunk 11]
```diff
@@ -150,6 +150,65 @@ public override void Remove(User entity)
...
+            }
+
+            this.Context.Entry(accessRequest).CurrentValues.SetValues(entity);
+            accessRequest.UpdatedById = this.User.GetUserId();
+            accessRequest.UpdatedOn = DateTime.UtcNow;
+            this.Context.Set<AccessRequest>().Update(accessRequest);
```
Comment (Fosol) [inline backend/dal/Services/Admin/Concrete/UserService.cs]: I'd change to `this.Context.AccessRequests.Update(accessRequest)`.  No need to use the generic method here, just adds additional reflection calls within EF.

File: backend/dal/Services/Admin/Concrete/UserService.cs [hunk 12]
```diff
@@ -167,12 +226,10 @@ public Paged<AccessRequest> GetAccessRequestsNoTracking(int page = 1, int quanti
                 .Include(p => p.Roles)
                 .ThenInclude(p => p.Role)
                 .Include(p => p.User)
-                .AsNoTracking();
+                .AsNoTracking()
+                .Where(ar => ar.User != null); // Access Requests with no user cannot be granted.
```
Comment (Fosol) [inline backend/dal/Services/Admin/Concrete/UserService.cs]: We shouldn't be saving access requests without users.  Makes me wonder if the additional filter here is needed.

File: backend/dal/Services/Admin/Concrete/UserService.cs (hunk 2 above)
Comment (devinleighsmith) [reply backend/dal/Services/Admin/Concrete/UserService.cs]: doesn't work for primatives.

File: backend/dal/Services/Admin/Concrete/UserService.cs (hunk 5 above)
Comment (devinleighsmith) [reply backend/dal/Services/Admin/Concrete/UserService.cs]: I originally wasn't enforcing the non-null user, missed updating the trickle down logic after I started enforcing it.

**Pull request merged**